### PR TITLE
[3392] Prevents edge from passing through another node

### LINKED ISSUE #3392: Prevents edge from passing through another node
### Actual behavior
edge_paths_example_with_existing

### Expected behavior

edge_paths_example_with_smart_edge

### PR DESCRIPTION
**Pull request by frouene**
Bug: https://github.com/eclipse-sirius/sirius-web/issues/3392

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [x] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: CHANGELOG.adoc [hunk 1]
```diff
@@ -57,6 +57,7 @@ This will make the dependency to the old Sirius Web architecture more explicit.
 - Move to docker compose v2 for our GitHub Actions workflow.
 - https://github.com/eclipse-sirius/sirius-web/issues/3277[#3277] [gantt] Move to @ObeoNetwork/gantt-task-react 0.4.8 to benefit for enhancements
 - [test] Switch to ArchUnit 1.2.1 to restore architectural tests which were failing recently
+- https://github.com/eclipse-sirius/sirius-web/issues/3392[#3392] [diagrams] Add dependence to @tisoap/react-flow-smart-edge 3.0.0
```
Comment (pcdavid) [inline CHANGELOG.adoc]: Typo: "dependenc[y]"

File: CHANGELOG.adoc [hunk 2]
```diff
@@ -110,7 +111,7 @@ They still support returning an `java.time.Instant` object directly.
 - https://github.com/eclipse-sirius/sirius-web/issues/3376[#3376] [diagram] Prevent `getToolSection` to be called during multi-selection
 - https://github.com/eclipse-sirius/sirius-web/issues/3377[#3377] [diagram] Only synchronize selection with explorer at the end of the process
 - https://github.com/eclipse-sirius/sirius-web/issues/3359[#3359] Add report on document upload
-
+- https://github.com/eclipse-sirius/sirius-web/issues/3392[#3392] [diagram] Prevents edge from passing through another node
```
Comment (pcdavid) [inline CHANGELOG.adoc]: Typo: "prevent" (no "s")

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/layout/layoutNode.ts [hunk 3]
```diff
@@ -475,12 +475,25 @@ export const applyRatioOnNewNodeSizeValue = (node: Node<NodeData>) => {
 export const isDescendantOf = (
   parent: Node,
   candidate: Node,
-  nodeById: (arg0: string | undefined) => Node | undefined
+  nodeById: (arg0: string) => Node | undefined
```
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams/src/renderer/layout/layoutNode.ts]: maybe give a proper name instead of "arg0"

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/layout/layoutNode.ts [hunk 4]
```diff
@@ -475,12 +475,25 @@ export const applyRatioOnNewNodeSizeValue = (node: Node<NodeData>) => {
...
-    const candidateParent: Node | undefined = nodeById(candidate.parentNode);
+    const candidateParent: Node | undefined = candidate.parentNode ? nodeById(candidate.parentNode) : undefined;
     return candidateParent !== undefined && isDescendantOf(parent, candidateParent, nodeById);
   }
 };
+
+export const isDescendantOfId = (
+  parent: string,
+  candidate: Node,
+  nodeById: (arg0: string) => Node | undefined
```
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams/src/renderer/layout/layoutNode.ts]: maybe give a proper name instead of "arg0"

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/layout/layoutNode.ts [hunk 5]
```diff
@@ -475,12 +475,25 @@ export const applyRatioOnNewNodeSizeValue = (node: Node<NodeData>) => {
...
     return true;
   } else {
-    const candidateParent: Node | undefined = nodeById(candidate.parentNode);
+    const candidateParent: Node | undefined = candidate.parentNode ? nodeById(candidate.parentNode) : undefined;
     return candidateParent !== undefined && isDescendantOf(parent, candidateParent, nodeById);
   }
 };
+
+export const isDescendantOfId = (
+  parent: string,
```
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams/src/renderer/layout/layoutNode.ts]: parentId?

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/layout/layoutNode.ts
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams/src/renderer/layout/layoutNode.ts:1]: Couldn't `isDescendantOf` be rewritenn as just:
```
export const isDescendantOf = (
  parent: Node,
  candidate: Node,
  nodeById: (arg0: string) => Node | undefined {
  return isDescendantOfId(parent.id, candidate, nodeById);
}
```
?

Comment (frouene) [reply packages/diagrams/frontend/sirius-components-diagrams/src/renderer/layout/layoutNode.ts:1]: Yeah, you're right
I changed it

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/edge/CommonEdge.tsx [hunk 6]
```diff
@@ -0,0 +1,113 @@
+/*******************************************************************************
```
Comment (mcharfadi) [inline packages/diagrams/frontend/sirius-components-diagrams/src/renderer/edge/CommonEdge.tsx]: I feel like CommonEdge should still be MultiLabelEdge and that your current MultiLabelEdge  should be called something like SmartEdgeWrapper or EdgeWrapper since it is only used to calculate the props passed to CommonEdge.

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/edge/CommonEdge.types.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Obeo - initial API and implementation
+ *******************************************************************************/
+
+export interface CommonEdgeProps {
```
Comment (mcharfadi) [inline packages/diagrams/frontend/sirius-components-diagrams/src/renderer/edge/CommonEdge.types.ts]: I propose something like
`
export type CommonEdgeProps<T = any> = {
  edgeCenterX: number;
  edgeCenterY: number;
  svgPathString: string;
} & EdgeProps<T>;
` to still use EdgeProps, what do you think ?

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/edge/EdgeLayout.ts [hunk 8]
```diff
@@ -22,7 +22,7 @@ import {
   GetUpdatedConnectionHandlesParameters,
   NodeCenter,
 } from './EdgeLayout.types';
-import { isDescendantOf } from '../layout/layoutNode';
+import { isDescendantOf, isDescendantOfId } from '../layout/layoutNode';
```
Comment (mcharfadi) [inline packages/diagrams/frontend/sirius-components-diagrams/src/renderer/edge/EdgeLayout.ts:25]: why did you need a new isDescendantOfId ?

**Review by mcharfadi — CHANGES_REQUESTED** [review]
This works fine and really improve the readability of the diagram.

I found that sometimes a shorter path could be used, but since it has to do with the handle position it's out of scope, just pointing that out for a potential futur improvement. 
smartEdges1

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/edge/MultiLabelEdge.tsx [hunk 9]
```diff
@@ -62,114 +82,192 @@ export const MultiLabelEdge = memo(
     targetPosition,
     sourceHandleId,
     targetHandleId,
-  }: EdgeProps<MultiLabelEdgeData>) => {
-    const theme = useTheme();
+  } = props;
+  const nodes = useNodes<NodeData>();
```
Comment (mcharfadi) [inline packages/diagrams/frontend/sirius-components-diagrams/src/renderer/edge/MultiLabelEdge.tsx]: please use ReactFlowInstance.getNodes() if you can.

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/edge/CommonEdge.tsx (hunk 6 above)
Comment (frouene) [reply packages/diagrams/frontend/sirius-components-diagrams/src/renderer/edge/CommonEdge.tsx]: Yes, you're right.
I've renamed it to `EdgeWrapper`

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/edge/CommonEdge.types.ts (hunk 7 above)
Comment (frouene) [reply packages/diagrams/frontend/sirius-components-diagrams/src/renderer/edge/CommonEdge.types.ts]: Well better.
Done

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/edge/EdgeLayout.ts (hunk 8 above)
Comment (frouene) [reply packages/diagrams/frontend/sirius-components-diagrams/src/renderer/edge/EdgeLayout.ts:25]: I didn't want to retrieve the entire node just to use its id

File: packages/diagrams/frontend/sirius-components-diagrams/src/renderer/edge/MultiLabelEdge.tsx (hunk 9 above)
Comment (frouene) [reply packages/diagrams/frontend/sirius-components-diagrams/src/renderer/edge/MultiLabelEdge.tsx]: done

**Review by mcharfadi — APPROVED** [review]
